diff --git a/zzip/memdisk.c b/zzip/memdisk.c
index 007e7ce..de14497 100644
--- a/zzip/memdisk.c
+++ b/zzip/memdisk.c
@@ -173,6 +173,11 @@ zzip_mem_entry_new(ZZIP_DISK * disk, ZZIP_DISK_ENTRY * entry)
         return 0;               /* errno=ENOMEM; */
     ___ struct zzip_file_header *header =
         zzip_disk_entry_to_file_header(disk, entry);
+    if (! header)
+    {
+        free (item);
+        return 0; /* errno=EBADMSG; */
+    }
     /*  there is a number of duplicated information in the file header
      *  or the disk entry block. Theoretically some part may be missing
      *  that exists in the other, ... but we will prefer the disk entry.
diff --git a/zzip/mmapped.c b/zzip/mmapped.c
index f685a84..b117b64 100644
--- a/zzip/mmapped.c
+++ b/zzip/mmapped.c
@@ -284,20 +284,25 @@ zzip_disk_entry_to_data(ZZIP_DISK * disk, struct zzip_disk_entry * entry)
  * This function does half the job of => zzip_disk_entry_to_data where it
  * can augment with => zzip_file_header_to_data helper from format/fetch.h
  *
- * returns: pointer into disk->buffer or 0 on error (bad format).
+ * returns: pointer into disk->buffer or 0 on error (errno = EBADMSG).
  */
 struct zzip_file_header *
 zzip_disk_entry_to_file_header(ZZIP_DISK * disk, struct zzip_disk_entry *entry)
 {
-    zzip_byte_t *file_header =  /* (struct zzip_file_header*) */
-        (disk->buffer + zzip_disk_entry_fileoffset(entry));
-    if (disk->buffer > file_header || file_header >= disk->endbuf)
+    zzip_byte_t *const ptr = disk->buffer + zzip_disk_entry_fileoffset(entry);
+    if (disk->buffer > ptr || ptr >= disk->endbuf)
+    {
+        errno = EBADMSG;
         return 0;
-#   define as_file_header (struct zzip_file_header *) file_header
-    if (zzip_file_header_get_magic(as_file_header) != ZZIP_FILE_HEADER_MAGIC)
+    }
+    ___  struct zzip_file_header *file_header = (void *) ptr;
+    if (zzip_file_header_get_magic(file_header) != ZZIP_FILE_HEADER_MAGIC)
+    {
+        errno = EBADMSG;
         return 0;
-    return as_file_header;
-#   undef  as_file_header;
+    }
+    return file_header;
+    ____;
 }

 /** => zzip_disk_entry_to_data
